src/plugins/intel_cpu/src/node.cpp

```diff
@@ -316,6 +315,118 @@ void Node::selectPreferPrimitiveDescriptor(const std::vector<impl_desc_type>& pr
     selectPrimitiveDescriptorByIndex(0);
 }
 
+bool Node::checkScalarShape(const ov::PartialShape& pshape) {
+    int oneNum = 0;
+    int sz = static_cast<int>(pshape.size());
+    for (auto s : pshape) {
+        if (s.is_static() && s.get_length() == 1) {
+            oneNum++;
+        }
+    }
+    return oneNum >= sz - 1;
+}
+
+void Node::selectPreferPrimitiveDescriptorWithShape(const std::vector<impl_desc_type>& priority, bool ignoreConstInputs) {
+    auto loopEdges = [&](size_t i,
+                         const size_t inputNodesNum,
+                         const ov::intel_cpu::NodeDesc& supportedPrimitiveDesc,
+                         int& reorderLocalCostScore) {
+        for (size_t j = 0; j < inputNodesNum; j++) {
+            auto parentEdge = getParentEdgeAt(j);
+            auto parentPtr = parentEdge->getParent();
+
+            // We don't take into account constant edges since reorders on them will be executed on load network
+            // stage
+            if (ignoreConstInputs && j > 0 && parentPtr->isConstant()) {
+                continue;
+            }
+
+            auto parent_spd = parentPtr->getSelectedPrimitiveDescriptor();
+            if (parent_spd != nullptr && !parent_spd->getConfig().outConfs.empty()) {
+                int inNum = parentEdge->getInputNum();
+                if (inNum < 0 || inNum >= static_cast<int>(parent_spd->getConfig().outConfs.size())) {
+                    inNum = 0;
+                }
+                auto curDesc = supportedPrimitiveDesc.getConfig().inConfs[j].getMemDesc();
+                auto parentDesc = parent_spd->getConfig().outConfs[inNum].getMemDesc();
+
+                const bool isCompatible = curDesc->isCompatible(*parentDesc);
+                bool isScalarShape = checkScalarShape(curDesc->getShape().toPartialShape());
+                if (!isCompatible) {
+                    if (isScalarShape) {
+                        reorderLocalCostScore += 1;
+                    } else {
+                        reorderLocalCostScore +=
+                            ov::shape_size<ov::intel_cpu::VectorDims>(curDesc->getShape().getStaticDims());
+                    }
+                }
+
+                DEBUG_LOG(getName(), " pd[", i, "].inConfs[", j, "]"
+                          " is ", (isCompatible ? "compatible" : "not compatible"),
+                          " shape is ", (isScalarShape ? "scalar" : "not scalar"),
+                          " with parent ", parentPtr->getName(),
+                          " outConfs[", inNum, "], reorderLocalCostScore add to ", reorderLocalCostScore);
+            }
+        }
+    };
+
+    auto loopSPD = [&](const impl_desc_type type, int& selectedPrimitive, int& reorderCostScore) {
```

Comment:
`selectSPDwithType`

Also let's avoid passing selectedPrimitiveIdx by reference.
Just return it by value.
I believe there is no reason to pass `reorderCostScore` as a parameter.
It can be just a local variable of the lambda.

Reply:
Updated.